Networks & Services API & UI

mavenugo opened the pull request:
This PR brings in the API & UI support for `networks` & `services` resource provided by `libnetwork`.
- API for `networks` & `services` resource
- UI hook for `docker network` & `docker service`
- `docker run --publish-service=SERVICE[.NETWORK][.DRIVER] ...
- Initial version of docs from @dave-tucker. (Will be updated with more upto-date content)

mavenugo committed:
Hook network chain to libnetwork

Signed-off-by: Madhu Venugopal <madhu@docker.com>

mavenugo committed:
Register libnetwork API router with docker daemon

libnetwork serves APIs for /networks/ resources.
register it as a Sub-Router with docker daemon's Router

Signed-off-by: Madhu Venugopal <madhu@docker.com>

dave-tucker committed:
Initial Network Docs

Signed-off-by: Dave Tucker <dt@docker.com>

cpuguy83 commented on api/client/network.go:
```diff
@@ -0,0 +1,9 @@
+func (cli *DockerCli) CmdNetwork(args ...string) error {
+	nCli := nwclient.NewNetworkCli(cli.out, cli.err, nwclient.CallFunc(cli.call))
+	args = append([]string{"network"}, args...)
+	return nCli.Cmd("docker", args...)
```
What's the purpose of sending "docker" here?

mavenugo replied:
NetworkCli takes in the "chain" parameter for better --help reasons.
Since `libnetwork` is an ingredient that provides networking service to docker, but also it has its own independent tool named `dnet` (which does everything networking). The same CLI chain under `network` is reused for both the cases. Hence passing the actual chain the network UI belongs to helps with consistent and better --help output.

cpuguy83 replied:
Ah, ok... so you want "docker" to display in the help output.
Should this be `os.Args[0]` instead?

mavenugo replied:
> Ah, ok... so you want "docker" to display in the help output.

Yes.

> Should this be os.Args[0] instead?

Good point :-) . i will fix that.

cpuguy83 commented on daemon/daemon.go:
```diff
@@ -980,6 +985,10 @@ func initNetworkController(config *Config) (libnetwork.NetworkController, error)
 	return controller, nil
 }
 
+func (daemon *Daemon) NetworkApiRouter() func(w http.ResponseWriter, req *http.Request) {
```
This doesn't look right. Does this need writer and request? Above where this is called nothing is being sent.
Also this is importing `net/http` into daemon which doesn't seem right.

mavenugo replied:
Yes. this (writer and request) is something Gorilla mux expects for handling the APIs. (Please refer to the api/server/server.go).
The reason I decided to keep this part of Daemon is that I didnt feel it is correct to expose the networkController out of daemon. Hence added this helper method which provides the APIHandler for network APIs.

cpuguy83 replied:
But these vars don't look like there being used at all.

mavenugo replied:
these are not vars. But it is a return value. being used in `registerSubRouter` to register the http handler for the `/networks` path.

cpuguy83 replied:
Oh silly me...  getting cross-eyed here!

cpuguy83 replied:
Maybe instead of importing http here we can define this as a type on api/server and use that instead?

I don't like that we're importing http here in daemon.

mavenugo replied:
Its a circular dependency issue (server depends on daemon).
I faced with the exact same dilemma : Should we expose networkController to api/server (or) import net/http in daemon. I chose the later because exposing networkController outside of daemon is worse :-)

cpuguy83 replied:
Oh right, my grep failed me. `docker/daemon` imports server, not `daemon/daemon`.
We could setup networkController in `docker/daemon` instead... which is what we do for registry already.

mavenugo replied:
sure. But, importing net/http is purely incidental due to the API requirements of a dependent module and is not reflective of the daemon functionality in anyway. 

I dont have any strong opinion on this. If you still prefer to keep this under `docker/daemon`, i can move this function there. `net/http` will move along with it :)

calavera replied:
If we don't want to import `het/http` in the daemon, regardless it's a good idea or not, we can export `daemon.netController` as public and make the call to `nwapi.NewHTTPHandler(daemon.NetController)` in the server. The bonus point is that we would not need this function either.

mavenugo replied:
@calavera yes. that is an option. but we took a decision not to make netController as public unless absolutely required.

cpuguy83 commented on docs/sources/reference/api/docker_remote_api_v1.19.md:
```diff
@@ -1252,12 +1252,12 @@ or being killed.
 Query Parameters:
 
--   **dockerfile** - path within the build context to the Dockerfile. This is 
+-   **dockerfile** - path within the build context to the Dockerfile. This is
```
I _think_ the whitespaces in this file are intentional as it signifies a new line for the markdown renderer.

mavenugo replied:
@dave-tucker can you please handle this ?

dave-tucker replied:
I *hope* it doesn't - let's check with @moxiegirl - trailing spaces are evil!

thaJeztah replied:
Unfortunately, they're part of the Markdown syntax and can be used to force a line break. Some of them are intended, some are just left-overs, so it's always tricky to remove them automatically

dave-tucker replied:
Ok in that case, I'll simply revert the whitespace changes for now.

cpuguy83 commented on integration-cli/docker_api_network_test.go:
```diff
@@ -0,0 +1,71 @@
+)
+
+func isNetworkAvailable(c *check.C, name string) bool {
+	status, body, err := sockRequest("GET", fmt.Sprintf("/v%s/networks", api.APIVERSION), nil)
```
Is there some reason we are passing in an `api.APIVERSION` here? If this is not specified it should just default to whatever is defined here anyway.

mavenugo replied:
Correct. that is because of https://github.com/docker/libnetwork/pull/209. once this PR is resolved in libnetwork and vendored in, we don't need the version parameter.
But, it is not wrong to have the version parameter (since Docker supports both the base-urls and hence both must work). This being just the Test code, we should be okay. But we will vendor in the above fix soon.

thaJeztah commented:
(Not in docs review yet, but a heads-up to prevent having to rewrite twice :))

Is this planned to be part of the 1.7 release? If not, the API docs will have to be moved to the v1.20 version, once created.

Also, the new end points will need to be added to the "what's new" section of the API docs

cpuguy83 commented on runconfig/parse.go:
```diff
@@ -476,14 +481,10 @@ func parseKeyValueOpts(opts opts.ListOpts) ([]KeyValuePair, error) {
 		if len(parts) < 2 || parts[1] == "" {
 			return "", fmt.Errorf("invalid container format container:<name|id>")
 		}
-	default:
```
Looks like we need to add a validation for an empty string, which this default clause was catching.

mavenugo replied:
Thanks. will fix it.

mavenugo commented:
@thaJeztah Yes, this is planned to be part of 1.7 and I believe @dave-tucker is working with @moxiegirl on this.

thaJeztah commented:
@mavenugo alright, thanks for checking. I'll add it to the milestone as well then, to better keep track.

cpuguy83 commented:
I will have to test this out tomorrow, but this looks _really_ sweet.
Is this going to be experimental?

mavenugo commented:
@cpuguy83 Thanks. the basic `network` command is not experimental. But the `network service` endpoint is already tagged as experimental in libnetwork.
If you build docker with the newly introduced `experimental` tag, you can already see the service appearing as experimental.

thaJeztah commented:
@mavenugo just saw this issue; https://github.com/docker/docker/issues/11247 wondering if that's a scenario that's possible with these new features?

mavenugo commented:
@thaJeztah There is a long answer and a short answer. Let me get geeky and give the longer answer ;-). 
As you are aware, libnetwork provides the flexibility to the end-users through network drivers. These drivers provide an implementation for the network plumbing requirements. These drivers are not global in nature. But we can have 1 specific driver for a specific network that is created. Also we support `labels` to the CNM constructs (network & endpoints) to let the user provide any configuration parameters to the network driver of choice for that particular deployment.
With the above constructs, anything is possible and it depends on the network driver and the feature it supports. Docker doesn't restrict any such implementation & the only requirement for the driver is to honor the CNM apis defined in libnetwork.

So, the short answer to the question is `it depends` on the driver that the user choose for their particular need. 

This approach also helps docker not to be overloaded with unlimited deployment scenarios and let the ecosystem provide solutions for a variety of user needs.

thaJeztah commented:
@mavenugo thanks, sounds really good! I'll add that to my answer on the linked issue.

dave-tucker commented:
@mavenugo couple of things:

1) `network` is not listed in the help of `docker --help`

2) We *should* support using name and ID for network interchangeably. This is really for ubiquity with current docker commands - so that a workflow like this is still possible.

```bash
mynet=$(docker network create foo)
docker run -itd --net=$mynet busybox
```

We could also go one step further and make `name` an optional parameter in `docker network create`. This would default to using docker's name generator for network names.

mavenugo commented:
@dave-tucker Good suggestions. will fix them.

We originally designed the network name as optional parameter in create using `--name` option (just like container), but later decided to make it mandatory. Applying the `pets vs cattle` analogy, Network / Service are like pets which the users might want to name them. whereas containers are like cattle & hence it can live with ids. But if you have a strong opinion, we can certainly make this change.

shykes commented:
PLEASE DO NOT MERGE UNTIL WE COMPLETE THE UI DISCUSSION.

We have not yet resolved whether we want the user-facing terms to be "networks" or "services". Feel free to discuss here. But please do not railroad. Thx.

shykes commented:
Here is a strawman for a UI.

```
# Print the current service discovery scope
$ docker service scope
work-dev

# Print available service discovery scopes
$ docker service scopes
work-prod
work-staging
* work-dev
work-weekly-demo
weekend-project

# List published services in the active scope (here work-dev)
$ docker service list
db
frontend
backend

# List published services in an explicit scope
$ docker service list --scope work-prod
db
db_slave_1
db_slave_2
ssl_gateway
frontend
backend

# Export the service discovery data from the active scope as a "/etc/hosts" resolver config file
$ docker service export --format=etchosts
10.0.0.1  db
10.0.0.2  db_slave_1
10.0.0.3 db_slave_2
10.0.0.4 ssl_gateway
10.0.0.5 frontend
10.0.0.6 backend

# Publish a new service in a specific scope.
# If the service is already published, it is updated to point to the new backend.
# This can be used to orchestrate sophisticated update strategies.
$ docker service --scope work-dev publish db --backend $CONTAINER_ID --replace=true

# As a convenience, when running a new container, automatically publish it as a service
# Here "db.work-dev" means "service name db in the scope work-dev".
$ docker run --publish-service db.work-dev
```

shykes commented:
Here the concept of "active scope" should be a client-side one, I think. No need to burden the daemon with this. In other words, the Remote API should always require an explicit scope name.

shykes commented:
A possible variation is to not use the `--scope` flag, but instead require the `NAME.SCOPE` syntax everywhere.

Variation 1: `docker service publish db --scope work-dev`
Variation 2: `docker service publish db.work-dev`

shykes commented:
Another possible variation: in `publish`, make the ID of the backend container mandatory, instead of supporting the `--backend` flag. This would make commands less verbose.

Variation A:     `docker service publish db --backend 8d0114b6027e --scope work-dev`

Variation B:     `docker service publish 8d0114b6027e db --scope work-dev`
Variation B+2: `docker service publish 8d0114b6027e db.work-dev`

mavenugo commented:
@shykes thanks for sharing your thoughts on this important service construct.

Will it help if we modify this PR into a more simplified version that supports just the `docker run` UI as you suggested and open a different enhancement to introduce the top-level `docker service` & give it the attention & time it rightly deserve ?

>> # As a convenience, when running a new container, automatically publish it as a service
>> # Here "db.work-dev" means "service name db in the scope work-dev".
>> $ docker run --publish-service db.work-dev

This UI along with maybe an experimental `--network-driver` will make it consistent with the experimental `--volume-driver` for 1.7.0.

Since the back-end can have multiple endpoints attached to it on different scopes, it would make it this run option as a list : `docker run --publish-service db.work-dev --publish-service demo.work-weekly-demo`

one other aspect is the swappability requirement of the container back-end with the service endpoint. We might have to introduce another option to have the user control if the back-end container is swappable.

Ofcourse, I feel there is lot more to the `docker service` than this simplified option pack as described above. But we can use it as a start & maybe experimental for 1.7. Please share your thoughts.

squaremo commented:
May I point out that users do not necessarily want to use docker's idea of services to provide networking. In particular, if they are using docker with other tools. So while I appreciate the urge to provide an expressive user interface, I would be very wary of obscuring networking with a service abstraction -- i.e., making service commands (and `--publish-service`) the _only_ way to get networking set up.

tomdee commented:
Just to echo what @mavenugo says above, I think getting the user visible high-level concepts and UI right is very important.  In terms of immediate proposal for 1.7 we support @mavenugo suggestion of moving the network UI to be experimental though it would be great to keep the API in stable.

I also agree with what @squaremo says – that it is important that the UI is flexible enough for a broad range to tools to use, beyond just those supplied by Docker.  

In terms of the strawman UI suggested by @shykes, we would need to flesh this out so it covers details such as how drivers are loaded and the lifecycle of the various network concepts.  E.g. Moving services (endpoints) between containers and generally decoupling them from containers lifecycles.

(And of course we need to resolve the “service scope” vs “network” terminology question.  In Calico we generally advocate intent based concepts where possible, and service scope could be argued as being a more intent oriented concept than network, though there are a lot of folks who still think in terms of networks.  I think as the UI is fleshed out it will become more obvious which is the right fit.)

We (Project Calico) would be really keen to help contribute to getting this concept right in 1.8 so it can move from experimental to stable.

thockin commented:
I didn't read the PR, so I thank @shykes for the higher level overview :)  I have a few questions here.

Full disclosure for anyone who doesn't know me, I work on Kubernetes.

First, I want to agree with @squaremo that networking and services, while often coupled, are not the same problem.  I think that functionality like this deserves to be exposed as a series of principled layers that each build on the previous.

It looks like you're adopting the Kubernetes concepts of "namespaces" (you spelled it 'scopes') and "services" but with some significant changes, and I am curious if that was intentional or accidental?  Specifically:

1. Kubernetes namespaces encompass almost all API objects, not just services.  For example, containers (pods) running in one namespace can not see containers in another namespace at all.  This extends to services, replication controllers, secrets, network volume claims, and so on.  Currently we enforce this visibility at the API level, and not the network level, but that is in discussion, too.  Are you intending your scopes to only provide scoping/namespacing those services or do you anticipate extending it to all objects (e.g. containers)?

2. Kubernetes services are specifically a decoupling mechanism from backends, rather than a link-by-another-name to a specific other container.  In your sketch, you said `--backend` - is it possible for a service to have more than one backend in your model?

3. Kubernetes services allow powerful zero-downtime rollout strategies in part by nature of their multi-backend capabilities and decoupling from specific instances.  You listed a `--replace` flag - is that just to distinguish create from update, or is there more power you are envisioning?

4. Are the IPs you list for services real or virtual?  If they are real, that seems to imply that either a) IPs must be floating across hosts and static to containers (so I can move my service-provider container, but what identifies a container as having been moved rather than new?) or b) something needs to manage discovery and (re-)binding.  You mention /etc/hosts but that is not easily updated on the fly for containers. If they are virtual, who manages them and the routing required to resolve them?

estesp commented on docs/sources/userguide/dockerlinks.md:
```diff
@@ -331,6 +333,74 @@ allowing linked communication to continue.
+
+Ok, now lets see if they can communicate!
+
+    $ docker exec web ping 192.168.99.20
```
minor nit: your names are **web1** and **db1** above, but the ping examples use **web** and **db**

ibuildthecloud commented:
@shykes I really have to think about this.  I 100% agree that docker needs a concept of a service because compose implicitly introduces one.  Not sure yet how this relates to networking.  I will do my best to respond to this thread before the weekend.

jainvipin commented:
`network` vs `service`:
- `network` is a basic block of interlinking containers. It would be good to see that getting exposed natively without a tie into `service`.
- `service` concept is needed, and could build on top of `network` or alongside. On that note, I liked the original proposal from @mavenugo  where `service` was a sub-command of `network`.

I also agree with @squaremo, @tomdee and @thockin about not mandating networking to come only via services. For a high level (orchestration tool, like docker-compose) hiding networks, etc. is desirable, but I am not sure it should be the case for docker daemon's UI.

erikh commented:
+1 to the separation between network and service

erikh commented:
So, I thought of two concrete situations where this notion of services does not fit, but networks do. I'm running with some old notions of what the networking initiative / CNM entails, but hopefully this is relevant commentary.

Setting multicast aside entirely...

1. Many distributed databases (I'm thinking of thrift+cassandra in particular) will not work with this notion of services. They need pools of addresses, not just a hostname. If given an single endpoint and that dies, well...

2. Failover. If the failover is not IP based, this will require many applications to be restarted if the underlying IP changes as they will do one resolve (database connections in particular). I don't know if the notion of "IP stability" is still on the roadmap but it is not here yet, as far as I know. If we had networks, we could use virtual IPs and CARP and other familiar systems on a bare metal network to provide this functionality.

shettyg commented:
For an Open vSwitch driver (for the OVN project), I like being able to provide a 'service' as an input to 'docker run'. It looks like many other users of Docker would like to provide 'network' as an input to 'docker run' as well. So would it be too bad to allow both? libnetwork as currently designed does handle both.

So you effectively can have both of the following:

```
docker run --publish-service db.work-dev:driverID
```

as well as:

```
docker run --net=:mynet:driverID 
```

shykes commented:
@thockin actually the main inspiration is the Dotcloud concept of "stacks" circa 2010. What is old is new again :) I think none of these concepts are groundbreaking, it's mostly a matter of exposing clean and useful anstractions to programmers and operators.

I don't think it's possible for everyone to agree on a single, strict definition of "service". But we can look at what is already a mainstream definition, reduce that to a sane and useful subset, and build on top of that. That way we avoid swimming against the current of mainstream understanding, and we avoid polarizing debates on technicalities that don't matter.

What is a decent mainstream definition of "service"? Something along the lines of "A standalone software component with which my application can communicate over the network to do something useful, without having to know how it's built or how it runs". What is a sane and useful subset of that? Perhaps "A network entrypoint which can be backed by different containers over time, allowing them to be discovered at a stable DNS name, communicate via a stable IP address, and expose a controlled set of tcp and udp ports".

Does this cover every definition of "service" under the sun? No. Does it solve a hair-on-fire problem for a lot of Docker users, using words that will seem natural and easy to grasp? I think so.

thockin commented:
The base definition works form, I think, and it suggests answers to some but not all of my questions.

Are you intending your scopes to only provide scoping/namespacing those services or do you anticipate extending it to all objects (e.g. containers)?

Is it possible for a service to have more than one backend in your model?

You listed a --replace flag - is that just to distinguish create from update, or is there more power you are envisioning?

Are the IPs you list for services real or virtual? Your response here says "stable IP" which reads as either virtual IPs or that the fabric must support IP migration across nodes.  Is that right?  Can you provide more detail?

vieux commented on integration-cli/docker_cli_network_test.go:
```diff
@@ -0,0 +1,101 @@
+	}
+
+	out, _, err = runCommandWithOutput(exec.Command(dockerBinary, "run", "--name", "should_fail", "--link", "host_container:tester", "busybox", "true"))
+	if err == nil || !strings.Contains(out, "User-Defined --netoption can't be used with links") {
```
what is this `--netoption` I don't see it defined anywhere ?

mavenugo replied:
its supposed to be `--net option`, both here and the corresponding error `ErrConflictUserNetworkAndLinks`.

vieux commented on runconfig/parse.go:
```diff
@@ -18,6 +18,7 @@ var (
 	ErrConflictNetworkAndDns            = fmt.Errorf("Conflicting options: --dns and the network mode (--net)")
 	ErrConflictNetworkHostname          = fmt.Errorf("Conflicting options: -h and the network mode (--net)")
 	ErrConflictHostNetworkAndLinks      = fmt.Errorf("Conflicting options: --net=host can't be used with links. This would result in undefined behavior")
+	ErrConflictUserNetworkAndLinks      = fmt.Errorf("Conflicting options: User-Defined --netoption can't be used with links.")
```
@mavenugo ok, the typo is here too.

mavenugo replied:
Yes. thats y the test works :) . double-negative.

ibuildthecloud commented:
Compose already has a concept of a service and it's very basic and I like it.  A service is just a collection of containers from the same configuration template.  The closest analogy I can think of is Amazons auto scaling groups.

Services need to talk to other services.  Today we have links.  I also like links because it allows you to describe your application as a graph of components.  Links are limited in that they describe a directional one to one relationship.

Networks add the ability to describe a many to many bidirectional relationship that is needed by many applications.  I would actually prefer to think of networks as a new form of link.   Today you can link service B to A as name FOO.  Now with networks you can link service A to network X as name FOO.

So after all that, I think all we really need here is to do "docker run --link-net containerid:name".  That seems sufficiently powerful for the immediate need for networks.

i think the discussion of a services UI is a different topic all together.  Which should happen, but I don't think the services UI is a pre-req for networks.

lxpollitt commented:
I agree with @ibuildthecloud that enhancing the services UI is something that shouldn't block defining the basic network UI.

A common requirement for the "networks" concept that we hear from many in the community is to be able to deploy the same service or application in isolated dev, test and prod networks.  In this model it is highly desirable that the service names and other container attributes are the same in each network, so for example the same compose files can be used across all the networks but the containers / services / applications deployed are still isolated from each other.  (So links would exist within the context of a network, rather than networks being a new kind of link.)  I think it is important that the 1.7 UI allows users to easily set up this type of isolation at the network level.

Links provide a basic/limited topology description today that is used primarily for service discovery purposes at the moment.  Some limited network policy can also be derived from links, but links do not capture some common network policy requirements (such as whether a container of a particular type is allowed to access the internet).  I think capturing network policy in a simple developer oriented intent based model should be high up the list of priorities for 1.8, but for 1.7 I think the networks concept as a very course grain isolation scope is a step in the right direction.

bboreham commented:
@ibuildthecloud I'm not clear how you get from "link service A to network X as name FOO" to "docker run --link-net containerid:name".  What would A, X and FOO be, in this example?

ibuildthecloud commented:
Sorry @bboreham I was a bit terse and I see I had a fundamental typo (blame that on my phone :) ).  It should have been `docker run --link-net networkname:name`, I updated my original comment to fix that typo.

First off, lets ignore "service".  I think that's a separate discussion so let me change the statement to "link _container_ A to network X as name FOO".  So the equivalent docker run would be

   docker run --name A --link-net X:FOO

This means container A will be known as FOO in network X.  From a DNS perspective any other container in network X can just do `ping FOO` and it will work.

mavenugo commented:
Thanks for all the suggestions and feedback. Based on all the discussions, i will push an update to this PR with `network` as a top-level UI component that provides low-level construct for many-to-many bidirectional communication as described in CNM https://github.com/docker/libnetwork/blob/master/docs/design.md .
Also we all agree that `service` is a very useful concept that has its rightful place as suggested by @shykes. Using this PR, we will be introducing both under the `experimental` tag. We hope having both these constructs under experimental provides the right level of balance between the various asks.

bfirsh commented:
As an update for those following along... this is currently blocked by vendoring a more recent version of libnetwork: https://github.com/docker/docker/pull/13682

shykes commented:
@thockin 

> *The base definition works for me, I think, and it suggests answers to some but not all of my questions.*

Good to know :) Thanks for taking the time to ask them.

> *Are you intending your scopes to only provide scoping/namespacing those services or do you anticipate extending it to all objects (e.g. containers)?*

Currently we are focused on discovery and connectivity between "things that can speak IP", because there is overwhelming demand for that, and almost every application can use it.

But, if you have a use case in mind for mixing "things that speak IP" with other "things", all in the same discovery scope, I'm totally open to considering it.

> *You listed a --replace flag - is that just to distinguish create from update, or is there more power you are envisioning?*

It's to distinguish "fail if it already has a backend" from "replace the old backend". This is really useful because it allows a simple primitive which other tools could build on to implement more ambitious things like rolling upgrades, selective rollbacks etc.

> *Are the IPs you list for services real or virtual?*

At the discretion of the driver. The contract is that routing must work within the same network scope, and must NOT work outside of it. So it can map to underlay or overlay for example.

>  Your response here says "stable IP" which reads as either virtual IPs or that the fabric must support IP migration across nodes. Is that right? Can you provide more detail?

By "stable IP" I meant that, as far as the application is concerned, once a particular name resolves to a particular IP in a particular discovery scope, that IP should not change. That's because we don't want to rely on the container's DNS resolver implementation (which we don't control) to always honor TTLs, and therefore we would rather not depend on DNS for dynamic changes.

This doesn't mean that the infrastructure *must* support IP migration across nodes. But it does mean that *if* you want to move containers across nodes without breaking the networking model of the application, then yes the infrastructure (represented by the network driver) will need to preserve the meaning of that IP within that scope. Note that "within that scope" is important, since the same IP could be used in parallel to mean different things across different resolution scopes. Currently we don't enforce IP allocation so it's possible for different drivers (or different instances of the same driver) to use the same IP in different networks. This will work fine *until* you decide to connect the same container to both networks: then you risk a conflict. We are still wondering how aggressively we should prevent these conflicts between drivers.

bboreham commented:
When I first read this wording around "stable IP", I worried that it was outlawing things like round-robin DNS.

Would it be right to take it as "the for a service IP shouldn't change under an application for infrastructure reasons", while leaving open "if some part of the service has crashed then you may be able to contact it on a different IP, which you could get if you re-queried DNS" ?

mavenugo committed:
Introducing default network and kvstore daemon flags

This commit brings in the ability to specify a default network and its
corresponding driver as daemon flags. This helps in existing clients to
make use of newer networking features provided by libnetwork which
includes multi-host networking capabilities. Also, in order to enable
multi-host capabilities, the kvstore configs are enabled as daemon
flags.

Signed-off-by: Madhu Venugopal <madhu@docker.com>

mavenugo committed:
Support for --publish-service flag in docker run

This commit makes use of the CNM model supported by LibNetwork and
provides an ability to let a container to publish a specified service.
Behind the scenes, if a service with the given name doesnt exist, it is
automatically created on appropriate network and attach the container.

Signed-off-by: Madhu Venugopal <madhu@docker.com>

squaremo commented on daemon/container_linux.go:
```diff
@@ -737,27 +737,89 @@ func (container *Container) buildCreateEndpointOptions() ([]libnetwork.EndpointO
+	if mode.IsDefault() {
+		canCreateNetwork = true
+		if service != "" {
+			service, networkName, networkDriver = parseService(controller, service)
```
If I read this and `parseService` above accurately, a user must supply a "service name" to be able to join a network at `docker run` time. This seems like a rather befuddling conflation of networks and service discovery. What if a user wants a load of containers that don't need to have service names?

I don't see what was the problem with keeping networking and service discovery distinct (and supporting, say, `docker --net=foo ..`). It doesn't make sense to hide networking under service discovery, since users may not always want the latter.

lxpollitt replied:
I agree with @squaremo.  Services and networks should be distinct concepts.  Hiding the basic primitives of networking behind a higher level construct that some (but not all) people will want to use feels wrong to me.  Services as envisaged here are a good thing (and plenty of people will want to use them).  But there are people who will not, particularly those running at large scale.  So I do think there is a lot of value in including some equivalent of the originally discussed "docker run --net=foo ..." which can be used independently of services.

For example, what is the proposed approach for someone who for example wants to spin up 100 back-end containers for their application running at scale?  Do they need to give each one a unique service name?  And DNS gets filled up with an entry for every container?  

I¹m concerned that for people running at scale this is going to be very painful.  We have customers looking to deploy Calico on thousands of servers in their data center, who are looking seriously at containers for their workloads.  In a data center of that size you are talking 100,000 containers.  Giving each one a unique service name and an entry in DNS?  I¹m not so sure that's going to work well for them.

We also need to remember that there are some fairly major orchestration systems that orchestrate Docker today which have their own service discovery mechanisms.  We need to make sure the Docker API & UI is friendly to their needs too.

erikh replied:
This is what I (poorly) said earlier, and I back this position.

-Erik

> On Jun 16, 2015, at 10:09 AM, Alex Pollitt <notifications@github.com> wrote:
> 
> In daemon/container_linux.go <https://github.com/docker/docker/pull/13441#discussion_r32544949>:
> 
> >  	if container.Config.NetworkDisabled || mode.IsContainer() {
> >  		return nil
> >  	}
> >  
> > -	var err error
> > +	var networkDriver string
> > +	service := container.Config.PublishService
> > +	networkName := mode.NetworkName()
> > +	canCreateNetwork := false
> > +	if mode.IsDefault() {
> > +		canCreateNetwork = true
> > +		if service != "" {
> > +			service, networkName, networkDriver = parseService(controller, service)
> I agree with @squaremo <https://github.com/squaremo>. Services and networks should be distinct concepts. Hiding the basic primitives of networking behind a higher level construct that some (but not all) people will want to use feels wrong to me. Services as envisaged here are a good thing (and plenty of people will want to use them). But there are people who will not, particularly those running at large scale. So I do think there is a lot of value in including some equivalent of the originally discussed "docker run --net=foo ..." which can be used independently of services.
> 
> For example, what is the proposed approach for someone who for example wants to spin up 100 back-end containers for their application running at scale? Do they need to give each one a unique service name? And DNS gets filled up with an entry for every container?
> 
> I¹m concerned that for people running at scale this is going to be very painful. We have customers looking to deploy Calico on thousands of servers in their data center, who are looking seriously at containers for their workloads. In a data center of that size you are talking 100,000 containers. Giving each one a unique service name and an entry in DNS? I¹m not so sure that's going to work well for them.
> 
> We also need to remember that there are some fairly major orchestration systems that orchestrate Docker today which have their own service discovery mechanisms. We need to make sure the Docker API & UI is friendly to their needs too.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/docker/docker/pull/13441/files#r32544949>.
>

thockin commented:
@ibuildthecloud

> docker run --name A --link-net X:FOO

Does this presume to support multiple --link-net flags?

I don't want to always beat the same drum, but I want to contrast this with
what we're doing in Kubernetes.  Fundamentally, we reject the idea that the
identity of an instance of a container is meaningful.  Containers will die and
be re-started.  They will be updated over time.  They will experience
unavailability, whether through crashes, deadlocks, machine failures, whatever.
We don't give meaningful names to containers (pods in our case, but same
point).  We DO give meaningful names to Services - but a service is not a
container, a service is a front for a set (in the mathematical sense, i.e. the
empty set is still a service) of containers.  This is such a fundamental point
that I just need to keep repeating it.

This discussion has wandered a lot from Networks to Services.  I really don't
think they are the same thing, and it looks like there's general agreement.
Networks are a substrate that govern reachability (and so on) of IP endpoints.
Services are a construct at a higher level of abstraction.

@shykes

It's not clear to me whether the design discussion around scopes and service in
this issue is actually a design discussion or idle chatter.  Does it make sense
to bust that out to a separate issue?

>> Are you intending your scopes to only provide scoping/namespacing those
>> services or do you anticipate extending it to all objects (e.g. containers)?
>
> Currently we are focused on discovery and connectivity between "things that
> can speak IP", because there is overwhelming demand for that, and almost every
> application can use it.

To be pedantic (or maybe I misunderstand) a container is (usually) a thing that
can speak IP, but you "scopes" sketch does not cover it.  I'm mostly asking
because of the value this construct (specced differently) has brought to
Kubernetes, which would be nice to see in docker core: thinner abstractions
FTW.  If you're really going to introduce namespacing, please consider making it
a more significant piece of the identity than the existing names.  The way you
describe scopes hiding/exposing information should apply to containers and
volumes and anything else.

>> Your response here says "stable IP" which reads as either virtual IPs or that
>> the fabric must support IP migration across nodes. Is that right? Can you
>> provide more detail?
>
> By "stable IP" I meant that, as far as the application is concerned, once a
> particular name resolves to a particular IP in a particular discovery scope,
> that IP should not change

I am 100% in agreement with a general policy of "assume all DNS libraries are
broken", and that is largely correct by experience.  My concern is this:  If you
lay down this rule, you make it very hard to recover from node failure.
Consider cases where a priori routing needs to be done to nodes (e.g. GCE): each
node gets a range of IPs it can allocate from.  Consider that a named thing (be
that a container or a service) lands on machine X and gets IP X1.  When machine
X dies, we must move that named thing elsewhere.  Now we assign it to machine Y.
I can no longer give it IP X1 - that doesn't route to machine Y.

My point is that you either define virtual IPs which abstract where a service
REALLY exists or you impose severe limitations on the environment.  Or you allow
DNS to not be 100% stable IPs.  Kubernetes went with virtual IPs, and so far it
has been great modulo some implementation long-tail problems that we are ironing
out.

monadic commented:
I'd like to +1 what the last few commenters wrote.  It seems to me that a great merit of Docker is to support a wide range of design choices.  @thockin describes one set, that make sense given the constraints and opinions of his platform.  @lxpollitt describes another set, reflecting the needs of very large scale network deployments.  @squaremo (my co-worker on Weave) points out UX issues.  Must this new abstraction be so limiting as to make all three of our cases so difficult?

mavenugo commented:
@squaremo @lxpollitt @monadic @thockin I can certainly see your points.
If you agree this discussion of treating `--net` not as a network `mode` (as it is currently implemented) and change it as a user-defined network to a different issue and we can get other opinion on that.

I think the concept of `--publish-service` is valid and we should let that feature happen via this PR (which also introduces the services concept). This will help us separate the concerns and move forward without mixing the concepts and blocked on each other.

I will open an issue for that thread and point to this PR.

ibuildthecloud commented:
Can I suggest a nice gathering at DockerCon to discuss this?

We have two perspectives here, one is the higher level service abstraction and the other is a lower level network abstraction.  I think they can be complementary, but it will be an extremely long dialogue on github to get to a point of reconciling these models.

ravigadde commented:
+1 for the discussion

mavenugo commented:
@ibuildthecloud if am not wrong there is one scheduled. @dave-tucker.
But lets not make this PR a planning session for that meeting :-)

thockin commented:
I think an IRL discussion would be way better than hashing it out here.  We
can circle back with a summary of the discussion and (hopefully) decisions.

Maybe you (Madhu) can pick a timeslot and a meeting place and let the
relevant parties know?

On Tue, Jun 16, 2015 at 6:12 PM, Madhu Venugopal <notifications@github.com>
wrote:

> @ibuildthecloud <https://github.com/ibuildthecloud> if am not wrong there
> is one scheduled. @dave-tucker <https://github.com/dave-tucker>.
> But lets not make this PR a planning session for that meeting :-)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/13441#issuecomment-112617524>.
>

ibuildthecloud commented:
@mavenugo I'm confused by PR #13977.  Does this mean that `docker networks/--net` will go forward independent of this services discussion?  Basically we could have both  `docker networks/--net` and services?

mavenugo commented:
@ibuildthecloud no. PR #13977 was opened just to continue the specific discussion brought up on reusing `--net` to supply the user-defined network name. This is in addition to the `--publish-service` construct that is defined using this PR. Since it is a well-defined problem statement, we can continue to discuss and hash it out using #13977 without overloading this PR.

We introduced the CNM model more than a month ago and we are unable to make use of it due to this UI discussion. I feel that the current balance that we have with network and service UI along with `--publish-service` satisfies the requirements that @shykes brought up & I think is a good one. 
Since these are defined under `experimental`, I think we should move on with this approach and get feedback from the users. Also with the help of #13977, we will get more inputs and we can continue to evolve this UI and I think that is the exact purpose of having experimental channel.

ibuildthecloud commented:
@mavenugo Sorry, this is not really constructive, but this whole discussion confuses me quite a bit.  But if I understand correctly we are choosing between networks or services.  We are also saying that the CNM may be nullified?

mavenugo commented:
@ibuildthecloud please go over the docs commit : https://github.com/mavenugo/docker/commit/31bfd53b33c593eb3a1f4a31eeb1dadf7bde6905 and that should clear up the latest status of this PR. After reviewing the docs, if you can give give it a hands on trial, that would answer your own question.

thockin commented:
Given the debate, I think separating networks and services into different
PRs makes a lot of sense.

On Tue, Jun 16, 2015 at 10:03 PM, Madhu Venugopal <notifications@github.com>
wrote:

> @ibuildthecloud <https://github.com/ibuildthecloud> please go over the
> docs commit : mavenugo@31bfd53
> <https://github.com/mavenugo/docker/commit/31bfd53b33c593eb3a1f4a31eeb1dadf7bde6905>
> and that should clear up the latest status of this PR. After reviewing the
> docs, if you can give give it a hands on trial, that would answer your own
> question.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/13441#issuecomment-112652924>.
>

ibuildthecloud commented:
@mavenugo Thanks, that helped a lot.

Services is an abstraction above containers but from what I read from those docs it's basically one-to-one.  How do we envision this services concept evolving.  In the current form in this PR, it does not seem to be the proper use of the term service because effectively a service is just an IP and DNS name for a container.  So basically an endpoint in the CNM.

ibuildthecloud commented:
Something seems off too me.  The problem I see is that people looking to implement network drivers will most definitely come from a networking background (duh).  So I imagine we will see VXLAN, VLAN, OVS, NSX, etc.  All of these technologies give you a virtual L2 network.  Service is very high level concept.  Publishing and discovering services is an even higher level concept.

By saying you publish a service to a network seems like we are mixing layers.  I have this very high level concept that will in practice be tied to a L2 boundary.  Frankly, I lean much more towards not even exposing the concept of a network to the user.  Just talk about services and their relationship.  Networking can be implied from service relationships.

lxpollitt commented:
@ibuildthecloud I think there may have been some conflating of the layers, which is why separating the services and networks PRs seems like a sensible way forward.  It is disappointing that it looks like there is little chance of getting this resolved (and anything implemented) in time for 1.7.  But at least DockerCon will give those of us who can make it a chance to discuss face to face, which should help reach a shared understanding faster.

I'm part of the Calico team, and therefore "from a networking background", but our philosophy is that developers should not need to worry about low level networking constructs such as L2 segments.  They should just be able to express their connectivity requirements in high-level intents, such as "containers of type A need to talk to containers of type B with constraints C" (where a constraint might be only allowing access to a particular port, or a bandwidth constraint, or whatever).  We believe the networking layer should sort out how to map that intent into the underlying networking primitives specific to that particular network driver implementation.

Having said that, a common use case which we hear from customers, is to be able to support dev, test and prod "networks", that are isolated from each other, without the need to change container or app or service definitions between the networks.  If you consider a "network" in this context, then saying a service exists within a network makes a fair amount of sense I think.

Anyway, I think I'm falling into the trap of lengthy comments here, when face to face might progress faster.  @ibuildthecloud - will you be at DockerCon?  It would be great to sync up with you then.  I suspect that our thinking may be very similar on most of this once we unpick it!

squaremo commented:
> I think the concept of --publish-service is valid and we should let that feature happen via this PR (which also introduces the services concept). This will help us separate the concerns and move forward without mixing the concepts and blocked on each other.

If we are going to separate the concerns, wouldn't it make sense to put the (clearly more controversial) services aspect elsewhere, rather than stepping on the bit that seemingly everyone agrees we should have?

monadic commented:
To reiterate what @squaremo said a few days ago: "I don't see what was the problem with keeping networking and service discovery distinct (and supporting, say, docker --net=foo ..). It doesn't make sense to hide networking under service discovery, since users may not always want the latter."

ibuildthecloud commented:
@mavenugo In this proposal the concepts are focused on publishing services.  What about consumption of services.  If I publish db.foo does web need to publish web.foo to consume db in foo?

bfirsh commented:
First part of this, the network UI, has been merged here: https://github.com/docker/docker/pull/14023

thockin commented:
Are we having a meeting Next week?  When and where?

On Tue, Jun 16, 2015 at 10:08 PM, Tim Hockin <thockin@google.com> wrote:

> Given the debate, I think separating networks and services into different
> PRs makes a lot of sense.
>
> On Tue, Jun 16, 2015 at 10:03 PM, Madhu Venugopal <
> notifications@github.com> wrote:
>
>> @ibuildthecloud <https://github.com/ibuildthecloud> please go over the
>> docs commit : mavenugo@31bfd53
>> <https://github.com/mavenugo/docker/commit/31bfd53b33c593eb3a1f4a31eeb1dadf7bde6905>
>> and that should clear up the latest status of this PR. After reviewing the
>> docs, if you can give give it a hands on trial, that would answer your own
>> question.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/docker/docker/pull/13441#issuecomment-112652924>.
>>
>
>

monadic commented:
there is a BoF on Tuesday @thockin

mavenugo commented:
Thanks for all the comments and good discussion around this PR.
Based on the popular demand, we split this into 2 independent PRs #14023, #14051.
Closing this PR.

squaremo commented:
> Based on the popular demand, we split this into 2 independent PRs #14023, #14051.

I don't think you can appeal to popular demand here -- you've just done the thing people found objectionable in two steps instead of one.

What I was asking for is a general purpose networks (and endpoints) UI, before jumping into services. I don't think you can close this until _at least_ #13977 or its near equivalent is addressed.

As @thockin said,

> networking and services, while often coupled, are not the same problem. I think that functionality like this deserves to be exposed as a series of principled layers that each build on the previous.

mavenugo commented:
@squaremo the purpose of this PR is to bring in the UI components and that is addressed. Hence it is closed. #13977 is opened to discuss on the very specific topic that you are interested. Hence lets use that PR for that purpose.
